### ISSUE TITLE
returning mongoose object not working

### ISSUE THREAD

**Issue by nschloe** [Automattic/mongoose#3856]
Interestingly, when upgrading to 4.4.1 (from 4.3.x), I noticed that returning the mongoose object doesn't work anymore,
```js
import mongoose from 'mongoose';

export const init = co.wrap(function* _init() {
  yield mongoose.connect(
    config.mongodbUri,
    {
      server: {
        socketOptions: {
          keepAlive: 1,
        },
      },
      replset: {
        socketOptions: {
          keepAlive: 1,
        },
      },
    }
  );

  mongoose.connection.on('error', (err) => {
    process.exit(1);
  });

  return mongoose;
});
```
always returns `undefined`. Within this function, it's all fine; `mongoose` is the fat object it used to be. I'm puzzled as to why this may happen.

**Comment by vkarpov15** [comment]
Yeah this is an edge case related to the changes in dc9612a5af412dc6f66365945b9acc67cccd578b - in order to support `yield mongoose.connect()` without backwards breaking changes like #3847, we made the `mongoose` object itself into a pseudo-promise that resolves to undefined, which is why when you `return mongoose` from co, `.then()` will get an undefined value. Working on a workaround.

**Comment by vkarpov15** [comment]
Actually, unfortunately there is no good workaround :( The only idea I have is to create a new object that gets returned from `connect()`, but that breaks chaining. This current approach (mongoose object as pseudo-promise) seems to be the best way to minimize impact, and the above case has a simple workaround:

```javascript
co(function*() {
  return { res: mongoose };
}).then(v => console.log(v.res));
```

Do you have any better suggestions?

**Comment by nschloe** [comment]
> pseudo-promise

Generally, I would consider it bad design making something look like A, but not fully implementing it. It leads to surprises like this one. I do understand and value your rigor in keeping backwards-compatibility though. It's always annoying if a minor version upgrade brings down your backend. (Like here for us.)

I would say we leave this bug open, and come back to it whenever you feel more comfy with breaking backwards compatibility (perhaps in a 5.x release).

**Comment by hgcummings** [comment]
This just tripped me up as well, as I wasn't treating mongoose itself as thenable but was returning it in a promise. It's quite confusing behaviour, and not at all obvious what's going on. Here's a minimal repro case that I just created for submitting a new issue (before spotting this one):

### config.js
```
'use strict';

let mongoose = require('mongoose');
mongoose.Promise = Promise;

let db = mongoose.connection;
mongoose.connect(process.env.MONGODB_URL);
module.exports = new Promise(function(resolve, reject) {
    db.once('open', () => {
        console.log('resolved mongoose', typeof mongoose);
        resolve(mongoose);
    });
    db.on('error', reject);
});
```

### index.js
```
'use strict';

const assert = require('assert'); 

require('./config.js').then(mongoose => {
    console.log('received mongoose', mongoose);
    assert.ok(mongoose);
}).then(
    () => process.exit(),
    err => { console.log(err); process.exit(1) }
);
```

The config.js module is designed to return a promise that provides a mongoose object only when the database connection is ready. I don't think I'm doing anything particularly unusual here (although genuinely happy to be corrected on this!) but the outcome is very surprising, and doesn't make sense without a deep understanding of how mongoose is implemented. The fact that mongoose is a thenable that resolves to `undefined` is quite a gotcha (especially as it didn't used to work like this in the previous minor version).

The other GitHub issues referred to above suggest that `mongoose.connect()` should return a promise-like object, which suggests that the correct fix for me is just to replace my `config.js` with:
```
'use strict';

let mongoose = require('mongoose');
mongoose.Promise = Promise;
module.exports = mongoose.connect(process.env.MONGODB_URL); // Promise-like
```

However, this doesn't work either, and also exports a promise that resolves to undefined. Is this expected?

For now, I'm just returning another object containing the mongoose object (as discussed above). Could you clarify what will be the 'correct' approach long term? - Would the next major version expect me to treat `mongoose.connect()` as thenable, allow me to return the mongoose object in a promise, or be compatible with both approaches?

**Comment by andrenarchy** [comment]
I'm also affected by this. The behavior is maximally confusing and I'd definitely for a clean solution (maybe with 5.x?).

**Comment by vkarpov15** [comment]
Well if this many people are affected by this issue, I suppose there's no other option than to fix it. @hgcummings you're right, that is the exact behavior that OP reported, I just never thought that somebody would return the mongoose singleton in a promise - that seems like a weird design decision. Will fix ASAP.

**Comment by vkarpov15** [comment]
So now mongoose is no longer thenable, but `.connect()` and `.disconnect()` return a thenable wrapper around a mongoose object. This is almost not backwards breaking, the only remaining edge case would be `mongoose.connect().options = { test: 1 }` would no longer do the same thing in 4.4. We can wrap that in a getter/setter if necessary, but would rather not unless somebody really needs that behavior.

**Comment by hgcummings** [comment]
That's great, thanks for the quick response.

I suppose it is a little pointless for me to be passing the mongoose object around when it could just be `require`d again. I think it's just a stylistic choice to make the configuration module's API appear more functional. I prefer a contract of "provide a fully-configured object that's ready to use" over "produce some side-effects on a singleton".

Thanks for supporting this usage anyway. I'll test it out when 4.4.3 drops.

**Comment by vkarpov15** [comment]
@hgcummings that's a fair argument. Hopefully the above fix will be sufficient - I'll push 4.4.3 on monday.

**Comment by hgcummings** [comment]
I can confirm that this is fixed for me in 4.4.3. Thanks!
